### ISSUE TITLE
Phase 3: SWE-bench Verified evaluation

### ISSUE THREAD

**Issue by athitten** [NVIDIA-NeMo/Automodel#2893]
This is **outside AutoModel** today; set up a separate eval environment:

1. **Install SWE-bench harness** — [evaluation guide](<https://www.swebench.com/SWE-bench/guides/evaluation/>)
2. **Choose eval scaffold** — for comparability, use **mini-SWE-agent** (same as official LM leaderboard)
3. **Serve the fine-tuned model** — vLLM or SGLang with Gemma4 chat template + tool-call format matching CoderForge training scaffold
4. **Run on SWE-bench Verified** (500 instances):

```
# Conceptual flow:
# 1. Agent loop generates patches per instance
# 2. Write predictions.jsonl (instance_id, model_patch)
# 3. swebench.harness.run_evaluation --predictions_path ... --run_id gemma4_31b_coderforge_cp8
```

5. **Baselines to report:**
   * Base `google/gemma-4-31B-it` (no SFT)
   * SFT model trained with CP

     **Important caveats** (from CoderForge authors):
     * Eval scaffold must match training scaffold for best results — CoderForge used OpenHands v0.52.1 tools

---

*Mirrored from Linear* [AM-557](https://linear.app/nvidia/issue/AM-557/phase-3-swe-bench-verified-evaluation) *— one-time snapshot, not auto-synced.< />< />* *  <br>**Sub-issue of* NVIDIA-NeMo/Automodel#2890.

**Comment by athitten** [comment]
> _Migrated from Linear [AM-557](https://linear.app/nvidia/issue/AM-557) — originally posted 2026-06-30._

* Verified that vLLM loads and serves the 20 step CP ckpt and emits correctly-parsed native tool_calls (run_bash({"command":"ls -F"}), finish_reason=tool_calls).
* With --tool-call-parser gemma4, the model emits a valid native tool call - correct function (run_bash), well-formed JSON args ({"command":"ls -F"}), content=null, finish_reason=tool_calls, and stop_reason=106 (the gemma4 `<turn|>` terminator).

vllm output:

```
Resolved architecture: Gemma4ForConditionalGeneration
non-default args: enable_auto_tool_choice=True, tool_call_parser='gemma4',
  tensor_parallel_size=2, max_model_len=8192, gpu_memory_utilization=0.9,
  limit_mm_per_prompt={'image':0,'audio':0}, dtype=torch.bfloat16
Gemma4 heterogeneous head dims (head_dim=256, global_head_dim=512) -> FlashAttention 4

Model load / serving

Starting to load model .../gemma4_31b_cp8_64k/epoch_0_step_19/model/consolidated
Loading weights took 16.92 seconds        # 59 GB across TP=2
... torch.compile + CUDA-graph capture (~3 min, this dominated the 320s READY) ...
Available KV cache memory: 35.07 GiB
GPU KV cache size: 83,427 tokens   |   Max concurrency @8192 tokens: 10.18x
Started server, GET /v1/models 200 OK   (READY=1 after 320s)

Tool-call test - request & full response

Request: a user message ("List the files... use the run_bash tool") + one run_bash function tool, tool_choice: auto.

{
  "model": "gemma4cf",
  "choices": [{
    "message": {
      "role": "assistant",
      "content": null,
      "tool_calls": [{
        "id": "chatcmpl-tool-8290016d37358190",
        "type": "function",
        "function": { "name": "run_bash", "arguments": "{\"command\": \"ls -F\"}" }
      }]
    },
    "finish_reason": "tool_calls",
    "stop_reason": 106
  }],
  "usage": { "prompt_tokens": 78, "completion_tokens": 20, "total_tokens": 98 },
  "system_fingerprint": "vllm-0.24.0-tp2-..."
}
PARSED_TOOL_CALLS: [{... "function": {"name": "run_bash", "arguments": "{\"command\": \"ls -F\"}"}}]
CONTENT: None
FINISH_REASON: tool_calls
POST /v1/chat/completions 200 OK   ---->   === JOB DONE ===
```

So the native tool-calling path that both mini-SWE-agent and OpenHands depend on works for the checkpoint. Full log: /lustre/fsw/portfolios/coreai/users/athittenaman//coderforge_cache/vllm_toolcheck_13270144.log

**Comment by athitten** [comment]
> _Migrated from Linear [AM-557](https://linear.app/nvidia/issue/AM-557) — originally posted 2026-06-30._

Eval scaffold: Finetuning uses OpenHands which is 3 tools (execute_bash, str_replace_editor, finish) but if we use **mini-SWE-agent which is the official SWE-bench leaderboard's reference scaffold, it exposes one tool: bash.**

As mentioned in the issue Eval scaffold must match training scaffold for best results — CoderForge used OpenHands v0.52.1 tools, thats why the plan is to run both

**Comment by athitten** [comment]
> _Migrated from Linear [AM-557](https://linear.app/nvidia/issue/AM-557) — originally posted 2026-07-01._

Phase 3 pipeline works e2e with SWE-bench-lite. Used lite since swe-bench-verified was not available in my quota of swe-bench.

What works: Serve (vLLM, gemma4 tool-calls, 131072 ctx) → mini-SWE-agent (native tool-calling) → enroot per-instance repo environments for agent rollout (Docker-less) → real git patches → preds.json → sb-cli submission.

Dry-run on 5 Lite instances (SFT step-300): 2/5 produced real patches, including the correct astropy-12907 fix; 3/5 hit RepeatedFormatError (expected for a lightly-trained 300-step model).

Plan was to use swe-bench cloud via sb-cli for grading since SWE-bench grading requires running each instance's test suite in that repo's exact environment (/testbed at the right commit, its conda env, FAIL_TO_PASS/PASS_TO_PASS tests). The official harness (swebench.harness.run_evaluation) does this by spinning up a Docker container per instance. But our cluster CW_DFW does not have docker only enroot + pyxis. So the standard Docker-based local grader couldn't run here. And SWE-bench cloud grading dint work for my account so switched local in-enroot grading the same setup that is used for agent rollout.

**Comment by athitten** [comment]
> _Migrated from Linear [AM-557](https://linear.app/nvidia/issue/AM-557) — originally posted 2026-07-01._

The final eval flow that is end-to-end functional with zero external dependencies (using local enroot grading):

vLLM serve → mini-SWE-agent (native tool calls) → enroot per-instance environments → real patches → local enroot grading (no Docker, no cloud). Scripts under `examples/convergence/coderforge/eval/`

**Comment by athitten** [comment]
> _Migrated from Linear [AM-557](https://linear.app/nvidia/issue/AM-557) — originally posted 2026-07-01._

The local grader was verified to be correct by running on gold patches. All 5/5 passed confirming the local enroot based grader is correct.

The CP-trained Gemma4-31B 300 steps ckpt (~0.2b) was evaluated with this eval workflow and it produced 2 patches out of 5 correctly. Both astropy-12907, astropy-6938 were correct fixes, proving CP-trained Gemma4-31B fixes real SWE-bench bugs.

**Comment by athitten** [comment]
Eval number of CP-trained Gemma4-31B 300 steps ckpt (\~0.2b) on SWE-bench-lite with OpenHands scaffolding: ~~25.1%~~  27.5%(resolve rate, total 300 instances, 27.5% after the correctness harness fix)

Eval number with mini-swe-agent scaffolding which the default that SWE-bench-lite uses was lower. Hence switched to OpenHands (OH) scaffolding which is the what fine-tuning used.

However the 0.2b SFT with CP on Gemma4-31B-it model does not match the SWE-bench-lite score of base Gemma4-31B-it which is at 42%.

Investigations done to bridge/explain the gap:

* Masking sanity. Verified the SFT actually trained on the right tokens (assistant turns incl. tool-calls; stop token 106 supervised;) and the  system + user + tool-result tokens are masked.
* Also ran the eval on 0.1B ckpt to check the prior ckpt and resolve rate was 25.8% (base IT model 42% -> 0.10 B ckpt 25.8% -> 0.20 B ckpt 25.1 %which shows, more SFT won't recover it.
* SFT ckpt fails to terminate. It keeps editing and re-editing, never calls finish, and runs out of turns. The CoderForge training trajectories are long (\~50 turns) and finish is only \~2% of the tool calls in them. So light imitation SFT taught the model the dominant behavior "keep acting/editing" but under-learned the skill to recognizing completion and calling finish, hence leading lower eval score.

Further investigation showed that out of the 78/300 tasks that were completed (called 'finish'), \~62 produced the correct patch and resolved. Remaining produced no patch (i.e zero edit) until max_turns=100 and in some cases the correct patch was produced at turn x (x <100) but the task continued editing until max_turns and providing an incorrect patch at the end.

**Comment by athitten** [comment]
Switched to SFT on base gemma4 31B instead of -IT model.

Base Gemma4-31B (non-IT) CoderForge SFT summary

Training (v1)
- SFT'd the raw base model on CoderForge (lr 5e-6, warmup 60, cp8xdp16, 64k). Ran until step 277; last checkpoint step_224 (~0.15B tokens).
- Retrained base at lr 1e-5, 800 steps (~0.5B tokens), loss 0.34 to 0.159, 4 windows with clean auto-resume.

First eval (broken harness)
- Base model (raw hf_gemma4_31b): capped at 0/31 partial (a non-instruct base can't do agentic tool-calling; run was crawling toward 0).
- Step_799 ckpt eval: after some fixes described below

Diagnostic chain (this is where the real findings are)
- Ruled out the wrong causes: tokenizer is byte-identical to the IT model, and the training labels DO contain the tool-call special tokens. So it was not a tokenizer bug, or masking bug.
- Root cause 1 (serve rendering): the gemma4 chat template's add_generation_prompt injects a free-form thinking channel (<|channel>thought) that CoderForge never used. That shoved the model off-distribution so it monologued and never emitted a tool call. Fix: pass enable_thinking=True per request. Verified: 0/3 structured tool calls before, 3/3 after (P of the tool-call token ~0.9999).
- Root cause 2 (arg format): the base model falls back to its pretrained JSON prior for args ({{"command":"..."}}) instead of the gemma4 native <|"|> format, and vLLM's gemma4 parser mangles that. This did NOT improve across steps 300/400/600, so training won't fix it. Fix: serve without the gemma4 parser (NOPARSER) and parse the raw JSON args ourselves.
- Root cause 3 (multi-turn solving): after fixes 1 and 2, the model tool-calls correctly and runs real commands (correct greps, correct file targeting, turn-1 returncode-0 commands), but on turn 2+ it loops on repeated commands and never lands an edit.


Key conclusions
- The SFT demonstrably installed agentic behavior on a base model that had none: valid structured tool calls, think-first workflow, phased exploration, correct file/symbol targeting, successful turn-1 actions. That is a real process-level uplift; end-to-end resolve is gated by deep-turn token consolidation.

Happening now
- Longer training: extended the run from 800 to 1600 steps (resumes from step_799, ~1.0B tokens total) to consolidate the deep-turn tool-call tokens (~takes 16 hrs, four 4 hrs windows). The resumed learning rate is healthy; re-eval the 1200 and 1599 checkpoints with the improved harness once ready

**Comment by athitten** [comment]
1600 step (~1B token eval):

The base model with this much SFT, installs the agentic process (tool calls, think-first, exploration, correct targeting, turn-1 successes) but not deep-turn coherence for end-to-end resolve.

**Comment by athitten** [comment]
Result of context trim:

- Capping observations at 1200 chars did not help the 1600 steps ckpt still collapses at turn 2 (command: "viewemakers"), same as untrimmed. Turns 3+ are garbage

**Comment by athitten** [comment]
Switching to Qwen3-32b to try and reproduce the exact numbers shared on [https://huggingface.co/datasets/togethercomputer/CoderForge-Preview](https://huggingface.co/datasets/togethercomputer/CoderForge-Preview)

Step 1: Reproduce 23% eval with swe-bench-verified (WIP)
Step 2: Add support for Qwen3-32b in AM
Step 3: Finetune for 1 epoch (~5 steps)
Step 4: swe-bench-verified eval

**Comment by athitten** [comment]
## Step 1: Reproduce 23% eval with swe-bench-verified

- **Reproduced CoderForge's Qwen3-32B base baseline within ~5 pts: full SWE-bench Verified (500) = 18.2% (91/500)** vs their reported 23.0% — up from our earlier 10.8%, gaps explained below.
- **Biggest fix — thinking mode:** the earlier low number was thinking-mode-**off**, not the model. Turning on Qwen3 thinking (temp 0.6 / top_p 0.95 / top_k 20) roughly doubled resolve on a 50-instance A/B (12% → 24%) and lifted the full 500 from 10.8% → 18.2% (+68% relative); the run was healthy (84% patch-attempt, 404/500 finished).
- **Disabled static YaRN:** A/B on 50 instances showed no-YaRN ≈ YaRN (12% vs 10%, within noise), but static YaRN is documented to degrade short contexts and most SWE trajectories fit native 40K — so we run at native context (no-YaRN), removing that downside cleanly.
- **Fixed a grading-infra bug that had produced false 0% scores:** the local enroot grader was failing `enroot create` on ~40/50 instances because each SWE-bench image extracts ~2.7 GB / 42k files and the grade node's small node-local `/tmp` filled after ~10 (errors were silently swallowed). Fix = point `ENROOT_DATA_PATH` at a large disk with a `df` check at start, and always verify the `enroot-errs` count before trusting a result — re-grading on a big-disk node returned the true numbers.
- **Remaining ~5 pt gap is a scaffold ceiling, not a CP/capability issue:** our harness reimplements OpenHands v0.52.1 rather than the real runtime CoderForge used — so **18.2% is now the locked base reference to measure the CP-SFT uplift against.**

**Comment by athitten** [comment]
Note on our implementaion of OpenHands:
OpenHands v0.52.1 is a full agent framework - thousands of lines: an agent controller, a prompt/memory system, real tool implementations, and a Docker-based sandbox runtime that CoderForge used both to generate the SFT training trajectories and (per their setup) to evaluate.

Our oh3_run.py (820 lines) copies only the tool schemas - the JSON names/params of execute_bash, str_replace_editor, finish - verbatim from OpenHands 0.52.1, and wraps them in a hand-written agent loop over our enroot backend. The docstring says it outright: "gives the model the EXACT 3-tool surface... without the cost/risk of running the full OpenHands runtime inside each container."

Why we reimplemented instead of running the real thing: OpenHands' real runtime spins up a Docker sandbox per instance. This cluster has no Docker (enroot only). So to eval at all, we rebuilt the tool surface over enroot.

**Comment by athitten** [comment]
> 1600 step (~1B token eval):
> 
> The base model with this much SFT, installs the agentic process (tool calls, think-first, exploration, correct targeting, turn-1 successes) but not deep-turn coherence for end-to-end resolve.

| | Base gemma4-31B (raw, un-SFT'd) | 1600-step CoderForge SFT on Base gemma4-31B |
|---|---|---|
| Total turns | 37 (aborted) | 125 msgs (60 action turns) |
| Structured `tool_calls` | 0 / 33 | 60 / 60  |
| Turn-1 behavior | repeats a planning template | coherent `think`, correct root-cause |
| Intended actions | none | `view /testbed`, `ls -la /testbed`, `find *.py`, `cd /testbed` |
| Failure mode | `never_toolcalled` | args garbled by `\uXXXX` tokens -> exit 127 |
| Resolved | 0 | 0 |

### Base — never takes an action

The raw base emits **zero tool calls** and just repeats a planning checklist verbatim every single turn (turns 2, 4, 6, 8, 10 … all identical):

> Phase 1. READING: read the problem and reword it in clearer terms  1.1 If there are code or config snippets…  1.2 Highlight message errors, method names, variables…

33 turns of this, no `execute_bash`, no `str_replace_editor`, no `finish` — the harness aborts at turn 37 (`empty=never_toolcalled`). The base model has no agentic ability: it plans forever and never acts.

### SFT-1600 — real agentic behavior, correct targeting

Turn 1 — a clean, structured `think` call that correctly diagnoses the actual bug:

> "The problem is with `separability_matrix`… when nesting compound models, the separability calculation doesn't properly propagate through the hierarchy."

Then it drives the right exploration — **60/60 turns are structured OpenHands tool calls**, and the intended commands (garbage stripped) are exactly what a solver should do:

```
think              -> (correct root-cause reasoning)
str_replace_editor -> view /testbed
execute_bash       -> ls -la /testbed
execute_bash       -> find /testbed -name "*.py" | head -20
execute_bash       -> cd /testbed && ls -la
```

**Where it breaks:** 50/60 (83%) of those calls have `\uXXXX` escape-garbage tokens injected into the arg values — `ക്കാരിls -la /testbedക്കാരി`, `西哥find /testbed…西哥` — so bash gets `$'ക…ls'` → exit 127, and no edit lands.
